clvm/read_cache_lookup.py

```diff
@@ -0,0 +1,148 @@
+    these objects that no longer exist, so we reference-count them.
+    """
+
+    def __init__(self):
+        """
+        Create a new `ReadCacheLookup` object with just the null terminator
+        (ie. and empty list of objects).
+        """
+        self.root_hash = hash_blobs(b"\1")
+        self.read_stack = []
+        self.count = Counter()
+        self.parent_lookup = {}
+
+    def push(self, obj_hash):
+        """
+        Note that an object with the given hash has just been pushed to
+        the read stack, and update the lookups as appropriate.
+        """
+        # we add two new entries: the new root of the tree, and this object (by id)
+        # new_root: (obj_hash, old_root)
+        new_root_hash = hash_blobs(b"\2", obj_hash, self.root_hash)
+
+        self.read_stack.append((obj_hash, self.root_hash))
+
+        self.count.update([obj_hash, new_root_hash])
+
+        new_parent_to_old_root = (new_root_hash, 0)
+        self.parent_lookup.setdefault(obj_hash, list()).append(new_parent_to_old_root)
+
+        new_parent_to_id = (new_root_hash, 1)
+        self.parent_lookup.setdefault(self.root_hash, list()).append(new_parent_to_id)
+        self.root_hash = new_root_hash
+
+    def pop(self):
+        """
+        Note that the top object has just been popped from the read stack.
+        Return the 2-tuple of the child hashes.
+        """
+        item = self.read_stack.pop()
+        self.count[item[0]] -= 1
+        self.count[self.root_hash] -= 1
+        self.root_hash = item[1]
+        return item
+
+    def pop2_and_cons(self):
+        """
+        Note that a "pop-and-cons" operation has just happened. We remove
+        two objects, cons them together, and push the cons, updating
+        the internal look-ups as necessary.
+        """
+        # we remove two items: the right side of each left/right pair
+        right = self.pop()
+        left = self.pop()
+
+        self.count.update([left[0], right[0]])
+
+        new_root_hash = hash_blobs(b"\2", left[0], right[0])
+
+        self.parent_lookup.setdefault(left[0], list()).append((new_root_hash, 0))
+        self.parent_lookup.setdefault(right[0], list()).append((new_root_hash, 1))
+        self.push(new_root_hash)
+
+    def find_path(self, obj, serialized_length):
+        """
+        This function looks for a path from the root to a child node with a given hash
+        by using the read cache.
+        """
+        if serialized_length < 3:
+            return None
+
+        seen_ids = set()
+
+        max_bytes_for_path_encoding = serialized_length - 2
+        # 1 byte for 0xfe, 1 min byte for savings
+
+        max_path_length = max_bytes_for_path_encoding * 8 - 1
+        seen_ids.add(obj)
+
+        partial_paths = [(obj, [])]
+
+        while partial_paths:
+            new_partial_paths = []
+            for (node, path) in partial_paths:
+                if node == self.root_hash:
+                    path.reverse()
```

Comment:
This reversal only happens once per call. Using deque would happen many times per call, and I'm concerned that it's not as efficient (or as simple) as `list.append`.

Reply:
According to https://docs.python.org/3/library/collections.html#collections.deque

> Deques are a generalization of stacks and queues (the name is pronounced “deck” and is short for “double-ended queue”). Deques support thread-safe, memory efficient appends and pops from either side of the deque with approximately the same O(1) performance in either direction.

`deque.appendLeft` is O(1) so I think it is as efficient as `list.append`.
Maybe internally `deque` is implemented as a linked list.

Reply:
The reversing is also `O(1)`, so this algorithm is still `O(1)` just as it would be with `deque`. The code here (as all the code in the pure python wheel) is not intended to be as fast as possible (although I am surprised how competive it is compared to the rust implementation), but as simple and as readable as possible, so I think it's okay to keep it as a list, since it's a built-in that a casual reader is likely to recognize.